smolyay/grid.py

```diff
@@ -0,0 +1,396 @@
+                grid_points_indexes_ = (numpy.array(
+                    numpy.meshgrid(*level_composition_index))
+                    .T.reshape(-1, dimension))
+                if grid_points_indexes is None:
+                    grid_points_indexes = grid_points_indexes_
+                else:
+                    grid_points_indexes = numpy.concatenate(
+                        (grid_points_indexes,
+                         grid_points_indexes_), axis=0)
+        # make integer grids, levels, grid points and basis functions
+        grid_points_indexes = grid_points_indexes.tolist()
+        level_lookup = {point: level for level, points in
+                        enumerate(self._basis_set.levels) for point in points}
+        levels = numpy.array([level_lookup[index] for index in
+                              numpy.concatenate(grid_points_indexes)]
+                             ).reshape((len(grid_points_indexes),
+                                        dimension)).tolist()
+        grid_points = numpy.array(self._basis_set.points
+                                  )[numpy.array(grid_points_indexes)].tolist()
+        grid_points_basis = numpy.array(self._basis_set.basis_functions)[
+            numpy.array(grid_points_indexes)].tolist()
+
+        return NestedIndexGrid(grid_points_indexes, grid_points,
+                               grid_points_basis, levels)
+
+
+class TensorGridGenerator(IndexGridGenerator):
+    """Create grid points and functions for constructing a surrogate.
+
+    Depending on the dimensionality, points and basis functions provided
+    by :class:`BasisFunctionSet`, the :meth:`__call__` makes full tensor
+    grids.
+    :meth:`__call__`  generates the indexes of
+    grid points (integer grids) depending on dimensionality (full tensor grid).
+    Grid points and their functions are made by replacing the indexes with
+    corresponding point.
+    """
+
+    def __init__(self, basis_set):
+        super().__init__(basis_set)
+
+    def __call__(self, dimension):
+        """Make grid points and their corresponding basis functions.
+
+        Depending on the dimensionality, a set of grid points are generated
+        (integer grid points). Grid points of the basis function, and
+        their corresponding basis functions are then generated by replacing
+        the indexes (integer numbers) with their points and functions.
+
+        Parameters
+        ----------
+        dimension: int
+            Number of independent variables.
+
+        Returns
+        -------
+        :class:`IndexGrid`
+            Grid points (and their indexes),their corresponding functions.
+        """
+        points_indexes = numpy.arange(len(self._basis_set.points))
+        # make integer grids, grid points and basis functions
+        grid_points_indexes = list(itertools.product(
+            *[points_indexes for point in range(dimension)]))
+        grid_points_indexes = list(map(list, grid_points_indexes))
```

Comment:
What is going on with list and map here? Is it because you need to make a list of lists?

Reply:
Yes. the list(itertools.product) generates list of tuples.